tools/apiview/parsers/cpp-api-parser/ApiViewProcessor/AstNode.cpp

```diff
@@ -1703,6 +1831,13 @@ class AstFunction : public AstNamedNode {
         m_exceptionSpecification{func->getExceptionSpecType()}
   {
     m_navigationId = GetNavigationId(func);
+
+    // If the type of the return value is in the global namespace, then flag it as an error.
+    if (m_returnValue.IsTypeInGlobalNamespace())
+    {
+      database->CreateApiViewMessage(ApiViewMessages::TypedefInGlobalNamespace, m_navigationId);
+    }
```

Comment:
Shame this method has to be manually added to every overload.  Could this be on the base?

Reply:
The base doesn't know what types are associated with a node - for instance, some nodes (like method nodes) have types, others (like TemplateTemplateParameter (which represents a template parameter to a template)) do not. So there's really no way of dropping it into the base class. 

Also the navigationId (which lets ApiView know where to put the error message) is top level node specific. And the database pointer is only used during the construction of the AstNode so it isn't included in the base class.

Putting that all together, if you were to put in a base class it would end up looking like:

```cpp
CheckForTypeInGlobalNamespace(database, m_returnValue, m_navigationId);
```

instead of:
```cpp
if (m_returnValue.IsTypeInGlobalNamespace())
{
  database->CreateApiViewMessage(ApiViewMessages::TypedefInGlobalNamespace, m_navigationId);
}
```

and I'm not sure it's a significant improvement.